[receiver/awsfirehose] Add new AWS Kinesis Data Firehose Receiver

jefchien opened the pull request:
**Description:** Made a receiver based on the AWS Firehose [documentation](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html) for the delivery endpoint and the AWS CloudWatch Metric Streams JSON format [documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats-json.html). Has a similar endpoint set up like the Zipkin receiver and encoding concept like the Kafka receiver. Maps the CloudWatch format:
```
{
    "metric_stream_name": "MyMetricStream",
    "account_id": "1234567890",
    "region": "us-east-1",
    "namespace": "AWS/EC2",
    "metric_name": "DiskWriteOps",
    "dimensions": {
        "InstanceId": "i-123456789012"
    },
    "timestamp": 1611929698000,
    "value": {
        "count": 3.0,
        "sum": 20.0,
        "max": 18.0,
        "min": 0.0
    },
    "unit": "Seconds"
}
```
to the `pdata.Metrics` format with each `metric_stream_name`/`account_id`/`region`/`namespace` set as a single `pdata.Resource` with one unspecified `pdata.InstrumentationLibrary` and each unique `metric_name` as a single `pdata.Metric`. The value is converted into a `pdata.SummaryDataPoint` with relevant `dimensions` added as attributes. All the metrics in a single request are aggregated into one `pdata.Metrics` and passed to the next consumer.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7571

**Testing:** <Describe what testing was performed and which tests were added.> Added unit tests for the receiver and the unmarshaler/builder.

**Documentation:** <Describe the documentation added.> Added README

**Sponsor:** @Aneurysm9 @alolita

jefchien committed:
Initial firehose receiver scaffolding.

jefchien committed:
Initial firehose receiver scaffolding.

jefchien committed:
Add firehose endpoint delivery request/response based on spec.

jefchien committed:
Add firehose endpoint delivery request/response based on spec.

jefchien committed:
Add CloudWatch Metric Stream encoding.

jefchien committed:
Add CloudWatch Metric Stream encoding.

jefchien committed:
Change unmarshaler to handle multiple records.

jefchien committed:
Change unmarshaler to handle multiple records.

jefchien committed:
Refactor receiver to make more generic.

jefchien committed:
Refactor receiver to make more generic.

jefchien committed:
Add unit tests.

jefchien committed:
Add unit tests.

jefchien committed:
Fix linting naming issues.

Add entry for versions.yaml and CHANGELOG.md.
Add common attributes to firehose consumer with tests.

jefchien committed:
Fix linting naming issues.

Add entry for versions.yaml and CHANGELOG.md.
Add common attributes to firehose consumer with tests.

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/unmarshaler.go:
```diff
@@ -0,0 +1,28 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unmarshaler // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/unmarshaler"
```
```suggestion
package unmarshaler // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/unmarshaler"
```

I would move this package hierarchy into an internal package to avoid unnecessarily increasing exposed API surface.

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/unmarshaler.go:
```diff
@@ -0,0 +1,97 @@
+				var metric cWMetric
+				err := json.Unmarshal(datum, &metric)
+				if err != nil || !u.isValid(metric) {
+					continue
```
Should this be logged?  I'd hate to lose visibility into why some metrics were not processed.

jefchien replied:
I'm not really sure how to log this. Would I log the entire `datum` as a string? This should only fail if the record isn't a CloudWatch metric. Like if the Firehose was for a different format. If all of the records in the request aren't CloudWatch metrics, then it'll return a `record format invalid` error. The fields it is checking will always be present in a metric.

Aneurysm9 replied:
If `err != nil` it is probably fine to do something like this:

```go
u.logger.Info("unable to unmarshal input", zap.Error(err), zap.Int("datum_index", di), zap.Int("row_index", ri))
```

Not sure how useful the index values would be, but if the user has access to the request it might help them pinpoint which particular point was failing.

One important consideration here is that the user on the sending side, with access to HTTP responses, may not be the same as the operator of the collector instance generating the responses.  Without logging on the collector side there will be no way for a collector operator to know that anything is going wrong or attempt to guide resolution.

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/unmarshaler.go:
```diff
@@ -0,0 +1,97 @@
+
+	md := pdata.NewMetrics()
+	for _, builder := range builders {
+		builder.Build().CopyTo(md.ResourceMetrics().AppendEmpty())
```
```suggestion
		builder.Build(md.ResourceMetrics().AppendEmpty())
```

Can the `CopyTo()` be avoided by providing the target directly to `Build()`?

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+// toMetricKey creates a key based on the metric name and dimensions to
+// keep metrics with different dimensions separate.
+func (rmb *resourceMetricsBuilder) toMetricKey(metric cWMetric) string {
+	return fmt.Sprintf("%s::%v", metric.MetricName, metric.Dimensions)
```
```suggestion
	return metric.MetricName
```

I think this should simply be keyed on metric name.  Different sets of dimensions for the metric can be represented as data points within the metric.

jefchien replied:
If they're in the same metric, wouldn't they be aggregated together? Based on the CloudWatch [documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Dimension), 

> Because dimensions are part of the unique identifier for a metric, whenever you add a unique name/value pair to one of your metrics, you are creating a new variation of that metric.

> CloudWatch treats each unique combination of dimensions as a separate metric, even if the metrics have the same metric name.

I believe we wouldn't want to aggregate metrics with the same name, but different dimensions since that could cause incorrect sums.

Aneurysm9 replied:
I think CloudWatch metrics are analogous to OTel's [timeseries model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#timeseries-model), whereas the pdata metric is of the [OTLP/stream model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#opentelemetry-protocol-data-model).

Here, we should probably map the CW metrics to [Summary data points](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#summary-legacy).  The metric level would contain just the [name, description, and unit](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L87), while the data point would [contain attributes, timestamps, and values](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L89-L90).

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+func (rmb *resourceMetricsBuilder) Build() pdata.ResourceMetrics {
+	rm := pdata.NewResourceMetrics()
+	ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	rmb.toResource().CopyTo(rm.Resource())
+	for _, mb := range rmb.metricBuilders {
+		mb.Build().CopyTo(ilm.Metrics().AppendEmpty())
```
```suggestion
	rmb.toResource(rm.Resource())
	for _, mb := range rmb.metricBuilders {
		mb.Build(ilm.Metrics().AppendEmpty())
```

By taking the target resource/metric directly we should be able to avoid copying.

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+
+// toResource creates a pdata.Resource from the fields in the resourceMetricsBuilder.
+// Splits the namespace into service.namespace/service.name if prepended by AWS/.
+func (rmb *resourceMetricsBuilder) toResource() pdata.Resource {
+	resource := pdata.NewResource()
```
```suggestion
func (rmb *resourceMetricsBuilder) toResource(resource pdata.Resource) {
```

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+// hasn't already been added
+func (mb *metricBuilder) AddDataPoint(metric cWMetric) {
+	if _, ok := mb.timestamps[metric.Timestamp]; !ok {
+		mb.toDataPoint(metric).CopyTo(mb.dataPoints.AppendEmpty())
```
```suggestion
		mb.toDataPoint(metric, mb.dataPoints.AppendEmpty())
```

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+
+// Build builds the pdata.Metric with the data points that were added
+// with AddDataPoint
+func (mb *metricBuilder) Build() pdata.Metric {
+	metric := pdata.NewMetric()
```
```suggestion
func (mb *metricBuilder) Build(metric pdata.Metric) {
```

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+	metric.SetUnit(mb.unit)
+	metric.SetDataType(pdata.MetricDataTypeHistogram)
+	metric.Histogram().SetAggregationTemporality(pdata.MetricAggregationTemporalityDelta)
+	mb.dataPoints.CopyTo(metric.Histogram().DataPoints())
```
```suggestion
	mb.dataPoints.MoveAndAppendTo(metric.Histogram().DataPoints())
```

This will move the data point slice into the new metric without copying/allocation.

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+
+// toDataPoint converts a cWMetric into a pdata datapoint and attaches the
+// dimensions as attributes
+func (mb *metricBuilder) toDataPoint(metric cWMetric) pdata.HistogramDataPoint {
+	dp := pdata.NewHistogramDataPoint()
```
```suggestion
func (mb *metricBuilder) toDataPoint(metric cWMetric, dp pdata.HistogramDataPoint) {
```

Aneurysm9 commented on receiver/awsfirehosereceiver/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,157 @@
+	dp := pdata.NewHistogramDataPoint()
+	dp.SetCount(uint64(metric.Value.Count))
+	dp.SetSum(metric.Value.Sum)
+	dp.SetExplicitBounds([]float64{metric.Value.Min, metric.Value.Max})
```
I'm not sure this is correct.  This would seem likely to result in bounds that differ for every data point in a timeseries.  I wonder whether it would be better to use a Summary with p0 and p100 instead.

jefchien replied:
Changed the metric data type to Summary.

jefchien commented on receiver/awsfirehosereceiver/receiver.go:
```diff
@@ -0,0 +1,255 @@
+// getReader uses the encoding to wrap the reader if matched otherwise returns
+// the original reader.
+func (fmr *firehoseReceiver) getReader(r io.Reader, encoding string) io.Reader {
+	if encoding == "gzip" {
```
Oh that's super useful. I was using the zipkin receiver as an example and they did the decompression as well. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/zipkinreceiver/trace_receiver.go#L165

jefchien commented on receiver/awsfirehosereceiver/Makefile:
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Added. It doesn't look like the other `Makefile`s have a new line EOF https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/awsxrayreceiver/Makefile

Aneurysm9 replied:
[POSIX has opinions](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html), and I share them. :)

>  3.403 Text File
> A file that contains characters organized into zero or more lines. ...

>  3.206 Line
> A sequence of zero or more non-\<newline\> characters plus a terminating \<newline\> character.

jefchien commented on receiver/awsfirehosereceiver/receiver.go:
```diff
@@ -0,0 +1,253 @@
+
+	statusCode, err := fmr.consumer.Consume(ctx, records, commonAttributes)
+	if err != nil {
+		fmr.sendResponse(w, requestID, statusCode, err)
```
Should it just log all the errors that are sent or is better to be selective?

Aneurysm9 replied:
Not sure.  Logging all errors at debug level or so is probably safest.  Maybe selectively make some info or warn if they indicate a more serious failure.

jefchien commented on receiver/awsfirehosereceiver/README.md:
```diff
@@ -0,0 +1,22 @@
+
+#### encoding:
+
+The encoding to use for the records received. Corresponds to an unmarshaler.
```
Yeah I agree that `encoding` is probably not the right term. I took the term from the Kafka receiver to try to keep it consistent. Can change it to something like `firehose_records_type`. Will update the README in general with more details.

jefchien commented on receiver/awsfirehosereceiver/config.go:
```diff
@@ -0,0 +1,27 @@
+	confighttp.HTTPServerSettings `mapstructure:",squash"`
+	Encoding                      string `mapstructure:"encoding"`
+	AccessKey                     string `mapstructure:"access_key"`
+}
```
If an unknown `Encoding` value is specified, then the receiver creation will return an error and fail. This way the correct unmarshaler set can be checked, so if a metrics `Encoding` value is specified for a trace pipeline, it will fail.

Aneurysm9 replied:
Checking in `Config.Validate()` will enable failing early as well as planned-but-not-implemented features like a config linter or dry run.  Better to die when parsing the config than waiting until a pipeline is constructed.

jefchien replied:
I'll probably have it die in both. Definitely need to make sure the encoding is going to work with the pipeline.

jefchien committed:
Move unmarshaler into internal.

Rename encoding to record_type. Change metrics builder to perform less copies. Add additional logging and errors. Add config validation.

jefchien committed:
Move unmarshaler into internal.

Rename encoding to record_type. Change metrics builder to perform less copies. Add additional logging and errors. Add config validation.

jefchien committed:
Add godocs. Update README with additional detail.

jefchien committed:
Add godocs. Update README with additional detail.

jefchien committed:
Change metrics data type to summary.

jefchien committed:
Change metrics data type to summary.

anuraaga commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,173 @@
+	attributes.InsertString(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+	attributes.InsertString(conventions.AttributeCloudAccountID, rmb.accountID)
+	attributes.InsertString(conventions.AttributeCloudRegion, rmb.region)
+	splitNamespace := strings.SplitN(rmb.namespace, namespaceDelimiter, 2)
```
Better to use `Index` and slices instead I think

jefchien replied:
Added `toServiceAttributes` function which uses `Index` and slices.

anuraaga commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/metricsbuilder.go:
```diff
@@ -0,0 +1,173 @@
+// toTimestampKey combines the dimensions and timestamps to create a key
+// used to prevent duplicate metrics.
+func (mb *metricBuilder) toTimestampKey(metric cWMetric) string {
+	return fmt.Sprintf("%v::%v", metric.Dimensions, metric.Timestamp)
```
Better to use a struct than a string for deduping, also we might need to sort the dimensions (other cloudwatch components use a `sorted.Label` lib from the otel sdk to handle that within a struct I think)

jefchien replied:
Added a `dataPointKey` struct with the timestamp and string dimensions. From [Go 1.12](https://go.dev/doc/go1.12#minor_library_changes), the fmt library prints dimensions in key-sorted order. Considered using `attributes.Distinct` instead, but involved a lot more allocation than seemed necessary and could not be reused for the `pdata.AttributeMap`.

```go
        var dimensions []attribute.KeyValue
	for k, v := range metric.Dimensions {
		dimensions = append(dimensions, attribute.String(k, v))
	}
	set := attribute.NewSetWithSortable(dimensions, sortable)
	return dataPointKey{
		timestamp: metric.Timestamp,
		dimensions: set.Equivalent(),
	}
```

anuraaga replied:
Thanks - I guess someday we'd want a pdata version of that

anuraaga commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/unmarshaler.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+// toResourceKey combines the metric stream name, namespace, account id, and region into a string key
+func (u Unmarshaler) toResourceKey(metric cWMetric) string {
```
Let's use a struct for this

jefchien replied:
Added a `resourceAttributes` struct with fields to use as key.

jefchien commented on receiver/awsfirehosereceiver/README.md:
```diff
@@ -0,0 +1,50 @@
+### record_type:
+The type of record being received from the delivery stream. Each unmarshaler handles a specific type, so the field allows the receiver to use the correct one.
+
+default: `cwmetrics`
```
The plan is to list them out in the `Record Types` section below the `Configuration` section. I'll point to it in the README.

jefchien committed:
Change string keys to structs.

Add additional test for receiver start up.

jefchien committed:
Change string keys to structs.

Add additional test for receiver start up.

jefchien commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/cwmetric.go:
```diff
@@ -0,0 +1,58 @@
+// The cWMetricValue is the actual values of the CloudWatch metric.
+type cWMetricValue struct {
+	// Max is the highest value observed.
+	Max float64 `json:"max"`
```
There isn't much documentation aside from https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats-json.html, but from my testing, these fields are always present even on user generated metrics.

jpkrohling committed:
update changelog

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jefchien committed:
Merge branch 'firehose-receiver' of github.com:jefchien/opentelemetry-collector-contrib into firehose-receiver

jpkrohling committed:
Merge branch 'main' into firehose-receiver

jefchien committed:
Bump up dependencies to v0.45.0.

jpkrohling commented on CHANGELOG.md:
```diff
@@ -66,6 +68,7 @@
 - `clickhouse` exporter: Add ClickHouse Exporter (#6907)
 - `pkg/translator/signalfx`: Extract signalfx to metrics conversion in a separate package (#7778)
   - Extract FromMetrics to SignalFx translator package (#7823)
+- `awsfirehose` receiver: Add AWS Kinesis Data Firehose Receiver (#7918)
```
I'll merge this as is, but this should be removed from v0.45.0.

jefchien replied:
Oh I thought it was.
